File: api/controllers/promotions.js
const { PrismaClient } = require('@prisma/client');
const prisma = new PrismaClient();

// GET /api/promotions - Get all promotions
const getPromotions = async (req, res) => {
  try {
    const promotions = await prisma.promotion.findMany({
      orderBy: { createdAt: 'desc' }
    });
    
    res.json(promotions);
  } catch (error) {
    console.log(error);
    res.status(500).json({ error: 'Failed to fetch promotions' });
  }
};

// GET /api/promotions/:id - Get promotion by ID
const getPromotionById = async (req, res) => {
  try {
    const { id } = req.params;
    
    const promotion = await prisma.promotion.findUnique({
      where: { id: parseInt(id) }
    });

    if (!promotion) {
      return res.status(404).json({ error: 'Promotion not found' });
    }

    res.json(promotion);
  } catch (error) {
    console.log(error);
    res.status(500).json({ error: 'Failed to fetch promotion' });
  }
};

// POST /api/promotions - Create new promotion
const createPromotion = async (req, res) => {
  try {
    const { 
      title, 
      description, 
      type, 
      value, 
      minDeposit, 
      maxBonus, 
      startDate, 
      endDate, 
      isActive, 
      terms 
    } = req.body;

    const promotion = await prisma.promotion.create({
      data: {
        title,
        description,
        type,
        value: parseFloat(value),
        minDeposit: minDeposit ? parseFloat(minDeposit) : null,
        maxBonus: maxBonus ? parseFloat(maxBonus) : null,
        startDate: new Date(startDate),
        endDate: new Date(endDate),
        isActive: isActive || false,
        terms
      }
    });

    res.status(201).json(promotion);
  } catch (error) {
    console.log(error);
    res.status(500).json({ error: 'Failed to create promotion' });
  }
};

// PUT /api/promotions/:id - Update promotion
const updatePromotion = async (req, res) => {
  try {
    const { id } = req.params;
    const { 
      title, 
      description, 
      type, 
      value, 
      minDeposit, 
      maxBonus, 
      startDate, 
      endDate, 
      isActive, 
      terms 
    } = req.body;

    const updateData = {};
    if (title) updateData.title = title;
    if (description) updateData.description = description;
    if (type) updateData.type = type;
    if (value) updateData.value = parseFloat(value);
    if (minDeposit) updateData.minDeposit = parseFloat(minDeposit);
    if (maxBonus) updateData.maxBonus = parseFloat(maxBonus);
    if (startDate) updateData.startDate = new Date(startDate);
    if (endDate) updateData.endDate = new Date(endDate);
    if (isActive !== undefined) updateData.isActive = isActive;
    if (terms) updateData.terms = terms;

    const promotion = await prisma.promotion.update({
      where: { id: parseInt(id) },
      data: updateData
    });

    res.json(promotion);
  } catch (error) {
    if (error.code === 'P2025') {
      return res.status(404).json({ error: 'Promotion not found' });
    }
    console.log(error);
    res.status(500).json({ error: 'Failed to update promotion' });
  }
};

// DELETE /api/promotions/:id - Delete promotion
const deletePromotion = async (req, res) => {
  try {
    const { id } = req.params;
    
    await prisma.promotion.delete({
      where: { id: parseInt(id) }
    });
    
    res.json({ message: 'Promotion deleted successfully' });
  } catch (error) {
    if (error.code === 'P2025') {
      return res.status(404).json({ error: 'Promotion not found' });
    }
    console.log(error);
    res.status(500).json({ error: 'Failed to delete promotion' });
  }
};

module.exports = {
  getPromotions,
  getPromotionById,
  createPromotion,
  updatePromotion,
  deletePromotion
};

File: api-client/promotions.py
import requests
import json
from typing import Dict, Any, List, Optional


def get_promotions(base_url: str) -> Dict[str, Any]:
    """Get all promotions"""
    url = f"{base_url}/api/v1/promotions"
    try:
        response = requests.get(url, headers={"Content-Type": "application/json"})
        response.raise_for_status()
        return {
            "success": True,
            "data": response.json(),
            "message": "Promotions fetched successfully"
        }
    except requests.exceptions.RequestException as e:
        return {
            "success": False,
            "error": f"Network error: {str(e)}",
            "status_code": getattr(e.response, 'status_code', None) if hasattr(e, 'response') else None
        }
    except Exception as e:
        return {
            "success": False,
            "error": f"Error: {str(e)}",
            "status_code": None
        }


def get_promotion_by_id(base_url: str, promotion_id: int) -> Dict[str, Any]:
    """Get promotion by ID"""
    url = f"{base_url}/api/v1/promotions/{promotion_id}"
    try:
        response = requests.get(url, headers={"Content-Type": "application/json"})
        response.raise_for_status()
        return {
            "success": True,
            "data": response.json(),
            "message": "Promotion fetched successfully"
        }
    except requests.exceptions.RequestException as e:
        return {
            "success": False,
            "error": f"Network error: {str(e)}",
            "status_code": getattr(e.response, 'status_code', None) if hasattr(e, 'response') else None
        }
    except Exception as e:
        return {
            "success": False,
            "error": f"Error: {str(e)}",
            "status_code": None
        }


def create_promotion(base_url: str, title: str, description: str, type: str, value: int, 
                    value_type: str, min_deposit: Optional[int] = None, max_bonus: Optional[int] = None,
                    code: Optional[str] = None, start_date: str = None, end_date: Optional[str] = None,
                    usage_limit: Optional[int] = None, player_usage_limit: Optional[int] = None,
                    is_active: bool = True) -> Dict[str, Any]:
    """Create new promotion"""
    url = f"{base_url}/api/v1/promotions"
    payload = {
        "title": title,
        "description": description,
        "type": type,
        "value": value,
        "valueType": value_type,
        "isActive": is_active,
        "startDate": start_date,
        "endDate": end_date
    }
    
    # Add optional fields if provided
    if min_deposit is not None:
        payload["minDeposit"] = min_deposit
    if max_bonus is not None:
        payload["maxBonus"] = max_bonus
    if code is not None:
        payload["code"] = code
    if start_date is not None:
        payload["startDate"] = start_date
    if end_date is not None:
        payload["endDate"] = end_date
    if usage_limit is not None:
        payload["usageLimit"] = usage_limit
    if player_usage_limit is not None:
        payload["playerUsageLimit"] = player_usage_limit

    try:
        response = requests.post(url, data=json.dumps(payload), headers={"Content-Type": "application/json"})
        response.raise_for_status()
        return {
            "success": True,
            "data": response.json(),
            "message": "Promotion created successfully"
        }
    except requests.exceptions.RequestException as e:
        return {
            "success": False,
            "error": f"Network error: {str(e)}",
            "status_code": getattr(e.response, 'status_code', None) if hasattr(e, 'response') else None
        }
    except Exception as e:
        return {
            "success": False,
            "error": f"Error: {str(e)}",
            "status_code": None
        }


def update_promotion(base_url: str, promotion_id: int, title: Optional[str] = None, 
                    description: Optional[str] = None, type: Optional[str] = None, 
                    value: Optional[int] = None, value_type: Optional[str] = None,
                    min_deposit: Optional[int] = None, max_bonus: Optional[int] = None,
                    code: Optional[str] = None, start_date: Optional[str] = None, 
                    end_date: Optional[str] = None, usage_limit: Optional[int] = None,
                    player_usage_limit: Optional[int] = None, is_active: Optional[bool] = None) -> Dict[str, Any]:
    """Update promotion"""
    url = f"{base_url}/api/v1/promotions/{promotion_id}"
    payload = {}
    
    # Only include fields that are provided
    if title is not None:
        payload["title"] = title
    if description is not None:
        payload["description"] = description
    if type is not None:
        payload["type"] = type
    if value is not None:
        payload["value"] = value
    if value_type is not None:
        payload["valueType"] = value_type
    if min_deposit is not None:
        payload["minDeposit"] = min_deposit
    if max_bonus is not None:
        payload["maxBonus"] = max_bonus
    if code is not None:
        payload["code"] = code
    if start_date is not None:
        payload["startDate"] = start_date
    if end_date is not None:
        payload["endDate"] = end_date
    if usage_limit is not None:
        payload["usageLimit"] = usage_limit
    if player_usage_limit is not None:
        payload["playerUsageLimit"] = player_usage_limit
    if is_active is not None:
        payload["isActive"] = is_active

    try:
        response = requests.put(url, data=json.dumps(payload), headers={"Content-Type": "application/json"})
        response.raise_for_status()
        return {
            "success": True,
            "data": response.json(),
            "message": "Promotion updated successfully"
        }
    except requests.exceptions.RequestException as e:
        return {
            "success": False,
            "error": f"Network error: {str(e)}",
            "status_code": getattr(e.response, 'status_code', None) if hasattr(e, 'response') else None
        }
    except Exception as e:
        return {
            "success": False,
            "error": f"Error: {str(e)}",
            "status_code": None
        }


def delete_promotion(base_url: str, promotion_id: int) -> Dict[str, Any]:
    """Delete promotion"""
    url = f"{base_url}/api/v1/promotions/{promotion_id}"
    try:
        response = requests.delete(url, headers={"Content-Type": "application/json"})
        response.raise_for_status()
        return {
            "success": True,
            "data": response.json() if response.content else None,
            "message": "Promotion deleted successfully"
        }
    except requests.exceptions.RequestException as e:
        return {
            "success": False,
            "error": f"Network error: {str(e)}",
            "status_code": getattr(e.response, 'status_code', None) if hasattr(e, 'response') else None
        }
    except Exception as e:
        return {
            "success": False,
            "error": f"Error: {str(e)}",
            "status_code": None
        }


if __name__ == "__main__":
    # Example usage
    base_url = "http://localhost:5000"
    
    # Get all promotions
    print("Getting all promotions:")
    promotions = get_promotions(base_url)
    print(promotions)
    
    # Create a new promotion
    print("\nCreating new promotion:")
    new_promotion = create_promotion(
        base_url=base_url,
        title="Welcome Bonus",
        description="100% deposit match up to $100",
        type="deposit_bonus",
        value=100,
        value_type="percentage",
        min_deposit=10,
        max_bonus=100,
        start_date="2025-08-27T00:00:00Z",
        end_date="2025-08-27T00:00:00Z"
    )
    print(new_promotion)
    
    # If creation was successful, test other operations
    if new_promotion["success"] and "data" in new_promotion:
        promotion_id = new_promotion["data"]["id"]
        
        # Get promotion by ID
        print(f"\nGetting promotion by ID {promotion_id}:")
        promotion = get_promotion_by_id(base_url, promotion_id)
        print(promotion)
        
        # Update promotion
        print(f"\nUpdating promotion {promotion_id}:")
        updated_promotion = update_promotion(
            base_url=base_url,
            promotion_id=promotion_id,
            description="Updated description: 100% deposit match up to $200",
            max_bonus=200
        )
        print(updated_promotion)
        
        # Delete promotion
        print(f"\nDeleting promotion {promotion_id}:")
        # deleted_promotion = delete_promotion(base_url, promotion_id)
        # print(deleted_promotion)
